src/lem/circuit.rs

```diff
@@ -484,13 +485,83 @@ struct RecursiveContext<'a, F: LurkField, C> {
     bindings: &'a VarMap<Val>,
 }
 
+fn synthesize_call<F: LurkField, CS: ConstraintSystem<F>, C: Coprocessor<F>>(
+    cs: &mut CS,
+    func: &Func,
+    not_dummy: &Boolean,
+    next_slot: &mut SlotsCounter,
+    bound_allocations: &mut BoundAllocations<F>,
+    ctx: &mut RecursiveContext<'_, F, C>,
+) -> Result<Vec<AllocatedPtr<F>>> {
+    let mut selected_branch = SelectedBranch {
+        selected_index: None,
+        branches: vec![],
+    };
+    synthesize_block(
+        cs,
+        &func.body,
+        &mut selected_branch,
+        not_dummy,
+        next_slot,
+        bound_allocations,
+        ctx,
+    )?;
+    Ok(allocate_return(cs, selected_branch))
+}
+
+/// The collection of all return values and `not_dummy`s of all
+/// branches in a block and the index of the uniquely selected
+/// return value, i.e. the one where the `not_dummy` is true.
+/// This index is only used to copy the correct values when
+/// allocating return variables, so that the constrains are
+/// satisfied.
+struct SelectedBranch<F: LurkField> {
+    selected_index: Option<usize>,
+    branches: Vec<(Boolean, Vec<AllocatedPtr<F>>)>,
+}
+
+/// Allocates variables for the return values and constrains them
+/// properly, given the collection of all return values for each
+/// branch. In the case where there is a unique branch, there is
+/// no need to allocate new variables.
+fn allocate_return<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    selected_branch: SelectedBranch<F>,
+) -> Vec<AllocatedPtr<F>> {
+    let SelectedBranch {
+        selected_index,
+        mut branches,
+    } = selected_branch;
+    assert!(!branches.is_empty());
+    if branches.len() == 1 {
+        let (_, output) = branches.pop().unwrap();
+        return output;
+    }
+    // If there is no selected branch, just choose whichever branch
+    let (_, selected_branch_output) = &branches[selected_index.unwrap_or(0)];
+    let output = selected_branch_output
+        .iter()
+        .enumerate()
+        .map(|(i, z)| {
+            let cs = ns!(cs, format!("matched output {i}"));
+            AllocatedPtr::alloc_infallible(cs, || z.get_value::<PtrTag>().unwrap())
+        })
+        .collect::<Vec<_>>();
+    for (branch_idx, (select, ptrs)) in branches.into_iter().enumerate() {
+        for (ptr_idx, (ptr, ret_ptr)) in ptrs.into_iter().zip(output.iter()).enumerate() {
+            ptr.implies_ptr_equal(ns!(cs, format!("{branch_idx}:{ptr_idx}")), &select, ret_ptr);
+        }
+    }
+    output
```

Comment:
The following might be a little bit better. It removes the `unwrap` inside the `alloc_infallible` closure.
```suggestion
) -> Result<Vec<AllocatedPtr<F>>> {
    let SelectedBranch {
        selected_index,
        mut branches,
    } = selected_branch;
    assert!(!branches.is_empty());
    if branches.len() == 1 {
        let (_, output) = branches.pop().unwrap();
        return Ok(output);
    }
    // If there is no selected branch, just choose whichever branch
    let (_, selected_branch_output) = &branches[selected_index.unwrap_or(0)];
    let mut output = Vec::with_capacity(selected_branch_output.len());
    for (i, z) in selected_branch_output.iter().enumerate() {
        let z_ptr = z.get_value::<PtrTag>().ok_or(anyhow!("Assignment missing for output {i}"))?;
        let ptr = AllocatedPtr::alloc_infallible(ns!(cs, format!("matched output {i}")), || z_ptr);
        output.push(ptr);
    }
    for (branch_idx, (select, ptrs)) in branches.into_iter().enumerate() {
        for (ptr_idx, (ptr, ret_ptr)) in ptrs.into_iter().zip(output.iter()).enumerate() {
            ptr.implies_ptr_equal(ns!(cs, format!("{branch_idx}:{ptr_idx}")), &select, ret_ptr);
        }
    }
    Ok(output)
```
Then `synthesize_call` can just end with `allocate_return(cs, selected_branch)` instead of wrapping with `Ok(...)`